lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -289,12 +292,24 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
     // stronger than this, checking also that the choice of lane/warp/block does
     // not affect the permutation of registers.  If we allow different
     // lane/warp/blocks to have different permutations, we can generalize this.
-    if (std::optional<LinearLayout> c = conversion.divideRight(
-            LinearLayout::identity1D(numLanes, kLane, kLane) *
-            LinearLayout::identity1D(numWarps, kWarp, kWarp) *
-            LinearLayout::identity1D(numBlocks, kBlock, kBlock));
-        c.has_value()) {
-      return transferWithinThread(*c, op, adaptor, rewriter);
+    if (cvtReordersRegisters(srcTy, dstTy)) {
+      // There are three possible cases:
+      //
+      // 1. `src_layout` has the same number of registers as `dst_layout`.
+      // 2. `src_layout` has fewer registers than `dst_layout`.
+      // 3. `src_layout` has more registers than `dst_layout`.
+      //
+      // In the second case, `conversion`, which is `src_layout .
+      // dst_layout^-1`, is not surjective because not all destination registers
+      // are covered.  Since the goal is to cover all of the destination
+      // registers, we can instead use the inverse conversion, which is
+      // `dst_layout . src_layout^-1`.
+      LinearLayout inverseConversion = dstLayout->invertAndCompose(*srcLayout);
+      auto dstToSrc = inverseConversion.divideRight(
+          LinearLayout::identity1D(numLanes, kLane, kLane) *
+          LinearLayout::identity1D(numWarps, kWarp, kWarp) *
+          LinearLayout::identity1D(numBlocks, kBlock, kBlock));
```

Comment:
I think it would be clearer to move this into the transferWithinThread function.

Fundamentally the current function creates a layout called `conversion`, which it uses to decide what kind of conversion to do -- reg-to-reg, shfl, shmem, whatever.  Then it calls transferWithinX, which figures out how to do the conversion.

If you moved this into the transferWithinThread function, you wouldn't need much of a comment at all, because you wouldn't be tempted to use `conversion`!  That is, it would be (more) obvious to do dstLayout->invertAndCompose(*srcLayout) rather than vice versa.

Reply:
(It's also confusing as-is that we take a variable called inverseConversion and pass it into a function where it's used in a variable called plain `conversion`.  Again moving the code into the function may clarify this...)